README.md
# HAProxy Exporter for Prometheus

This is a simple server that scrapes HAProxy stats and exports them via HTTP for
Prometheus consumption.

## This exporter is retired

In all supported versions of HAProxy, the official source includes a Prometheus exporter module that can be built into your binary with a single flag during build time and offers a native Prometheus endpoint. For more information see [down below](#official-prometheus-exporter).

Please transition to using the built-in support as soon as possible.

## Getting Started

To run it:

```bash
./haproxy_exporter [flags]
```

Help on flags:

```bash
./haproxy_exporter --help
```

For more information check the [source code documentation][gdocs]. All of the
core developers are accessible via the Prometheus Developers [mailinglist][].

[gdocs]: http://godoc.org/github.com/prometheus/haproxy_exporter
[mailinglist]: https://groups.google.com/forum/?fromgroups#!forum/prometheus-developers

## Usage

### HTTP stats URL

Specify custom URLs for the HAProxy stats port using the `--haproxy.scrape-uri`
flag. For example, if you have set `stats uri /baz`,

```bash
haproxy_exporter --haproxy.scrape-uri="http://localhost:5000/baz?stats;csv"
```

Or to scrape a remote host:

```bash
haproxy_exporter --haproxy.scrape-uri="http://haproxy.example.com/haproxy?stats;csv"
```

Note that the `;csv` is mandatory (and needs to be quoted).

If your stats port is protected by [basic auth][], add the credentials to the
scrape URL:

```bash
haproxy_exporter  --haproxy.scrape-uri="http://user:pass@haproxy.example.com/haproxy?stats;csv"
```

Alternatively, provide the password through a file, so that it does not appear in the process
table or in the output of the ```/debug/pprof/cmdline``` profiling service:

```bash
echo '--haproxy.scrape-uri=http://user:pass@haproxy.example.com/haproxy?stats;csv' > args
haproxy_exporter @args
```

You can also scrape HTTPS URLs. Certificate validation is enabled by default, but
you can disable it using the `--no-haproxy.ssl-verify` flag:

```bash
haproxy_exporter --no-haproxy.ssl-verify --haproxy.scrape-uri="https://haproxy.example.com/haproxy?stats;csv"
```

If scraping a remote HAProxy must be done via an HTTP proxy, you can enable reading of the
standard [`$http_proxy` / `$https_proxy` / `$no_proxy` environment variables](https://pkg.go.dev/net/http#ProxyFromEnvironment) by using the
`--http.proxy-from-env` flag (these variables will be ignored otherwise):

```bash
export HTTP_PROXY="http://proxy:3128"
haproxy_exporter --http.proxy-from-env --haproxy.scrape-uri="http://haproxy.example.com/haproxy?stats;csv"
```

[basic auth]: https://cbonte.github.io/haproxy-dconv/configuration-1.6.html#4-stats%20auth

### Unix Sockets

As alternative to localhost HTTP a stats socket can be used. Enable the stats
socket in HAProxy with for example:


    stats socket /run/haproxy/admin.sock mode 660 level admin


The scrape URL uses the 'unix:' scheme:

```bash
haproxy_exporter --haproxy.scrape-uri=unix:/run/haproxy/admin.sock
```

### Docker

[![Docker Repository on Quay](https://quay.io/repository/prometheus/haproxy-exporter/status)][quay]
[![Docker Pulls](https://img.shields.io/docker/pulls/prom/haproxy-exporter.svg?maxAge=604800)][hub]

To run the haproxy exporter as a Docker container, run:

```bash
docker run -p 9101:9101 quay.io/prometheus/haproxy-exporter:latest --haproxy.scrape-uri="http://user:pass@haproxy.example.com/haproxy?stats;csv"
```

[hub]: https://hub.docker.com/r/prom/haproxy-exporter/
[quay]: https://quay.io/repository/prometheus/haproxy-exporter

## Development

[![Go Report Card](https://goreportcard.com/badge/github.com/prometheus/haproxy_exporter)][goreportcard]
[![Code Climate](https://codeclimate.com/github/prometheus/haproxy_exporter/badges/gpa.svg)][codeclimate]

[goreportcard]: https://goreportcard.com/report/github.com/prometheus/haproxy_exporter
[codeclimate]: https://codeclimate.com/github/prometheus/haproxy_exporter

### Building

```bash
make build
```

### Testing

[![CircleCI](https://circleci.com/gh/prometheus/haproxy_exporter/tree/main.svg?style=shield)][circleci]

```bash
make test
```

[circleci]: https://circleci.com/gh/prometheus/haproxy_exporter

### TLS and basic authentication

The HAProxy Exporter supports TLS and basic authentication.

To use TLS and/or basic authentication, you need to pass a configuration file
using the `--web.config.file` parameter. The format of the file is described
[in the exporter-toolkit repository](https://github.com/prometheus/exporter-toolkit/blob/master/docs/web-configuration.md).

## License

Apache License 2.0, see [LICENSE](https://github.com/prometheus/haproxy_exporter/blob/main/LICENSE).

## Alternatives

### Official Prometheus exporter

As of 2.0.0, HAProxy includes a Prometheus exporter module that can be built into your binary during build time.
For HAProxy 2.4 and higher, pass the `USE_PROMEX` flag to `make`:

```bash
make TARGET=linux-glibc USE_PROMEX=1
```

Pre-built versions, including the [Docker image](https://hub.docker.com/_/haproxy), typically have this enabled already.

Once built, you can enable and configure the Prometheus endpoint from your `haproxy.cfg` file as a typical frontend:

```haproxy
frontend stats
    bind *:8404
    http-request use-service prometheus-exporter if { path /metrics }
    stats enable
    stats uri /stats
    stats refresh 10s
```

For more information, see [this official blog post](https://www.haproxy.com/blog/haproxy-exposes-a-prometheus-metrics-endpoint/).


CHANGELOG.md
## 0.15.0 / 2023-02-15

* [FEATURE] Add metric for idle time percentage #236 #255
* [ENHANCEMENT] Dependency updates #252 #253 #254

This is the **FINAL RELEASE** of the standalone HAProxy exporter.
All supported versions of HAProxy now have Prometheus metrics [built in](https://github.com/prometheus/haproxy_exporter#official-prometheus-exporter).
Please transition to using the built-in support as soon as possible.

## 0.14.0 / 2022-11-29

* [SECURITY] Update Exporter Toolkit (CVE-2022-46146) #251
* [FEATURE] Support multiple Listen Addresses and systemd socket activation #251

## 0.13.0 / 2021-11-26

* [FEATURE] Add TLS and Basic authentication #205
* [ENHANCEMENT] Added average over last 1024 requests metrics to server metric type #196
* [BUGFIX] Fix docker images architecture and publish ppc64le & s390x images #211

## 0.12.0 / 2020-12-09

* [ENHANCEMENT] Add --version flag #189
* [BUGFIX] Use newest Go version to fix random panic in the runtime
* [BUGFIX] Fix typos in log messages #188 #191

## 0.11.0 / 2020-06-21

* [CHANGE] Switch logging to go-kit #171
* [CHANGE] Fix metric types #182
* [CHANGE] Fix unit of time metric #183
* [FEATURE] Add filtering on server status #160
* [ENHANCEMENT] Add compression and server selection metrics #154
* [ENHANCEMENT] Add client/server abort metrics #167
* [ENHANCEMENT] Add version info metric (when using UNIX sockets) #180

Note: This release fixes the metric types of counters and renames the following metrics:

* `haproxy_exporter_csv_parse_failures` -> `haproxy_exporter_csv_parse_failures_total`
* `haproxy_exporter_total_scrapes` -> `haproxy_exporter_scrapes_total`
* `haproxy_server_check_duration_milliseconds` -> `haproxy_server_check_duration_seconds`

## 0.10.0 / 2019-01-15

* [ENHANCEMENT] Convert metrics collection to Const metrics #139
* [BUGFIX] Fix silent dropping of metrics for older versions of haproxy #139

## 0.9.0 / 2018-01-23

* [CHANGE] Rename `*_connections_total` to `*_sessions_total` following the changes in HAProxy 1.7
* [ENHANCEMENT] Add new `haproxy_server_connections_total` metric
* [ENHANCEMENT] Add new `--haproxy.ssl-verify` flag
* [BUGFIX] Convert latency metrics to correct unit.

## 0.8.0 / 2017-08-24

* [CHANGE] New flag handling (double dashs are required)
* [FEATURE] Add metric for session limit.
* [FEATURE] Add metrics for average HTTP request latency

## 0.7.1 / 2016-10-12

* [BUGFIX] Fix timeout behavior when reusing HTTP connections
* [BUGFIX] Remove usage of undocumented golang type assertion behavior

## 0.7.0 / 2016-06-08

* [FEATURE] Add support for unix sockets

## 0.6.0 / 2016-05-13

* [CHANGE] Use new build process, changes the structure of the tarball.
* [FIX] Abort on non-200 status code from HAProxy.
* [ENHANCEMENT] Add -version flag and version metric.
* [ENHANCEMENT] Add chkfail and downtime server metrics.
* [ENHANCEMENT] Remove locks and unnecessary channel communication.

## 0.5.2 / 2016-04-05

* [FIX] Limit graceful CSV error handling to parse errors

## 0.5.1 / 2016-03-31

* [FIX] Handle invalid CSV lines gracefully

## 0.5.0 / 2015-12-23

* [CHANGE] New Dockerfile
* [ENHANCEMENT] Export haproxy_check_duration_milliseconds
* [ENHANCEMENT] Export haproxy_limit_sessions
* [ENHANCEMENT] Export haproxy_limit_session_rate
* [ENHANCEMENT] Allow complete deactivation of server metrics
* [ENHANCEMENT] Use common prometheus logging
* [FIX] Fix status field parsing of servers in MAINT status


haproxy_exporter_test.go
// Copyright 2018 The Prometheus Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bufio"
	"io"
	"net"
	"net/http"
	"net/http/httptest"
	"os"
	"path"
	"reflect"
	"runtime"
	"testing"
	"time"

	"github.com/go-kit/log"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/testutil"
)

const (
	testSocket = "/tmp/haproxyexportertest.sock"
	testInfo   = "Release_date: test date\nVersion: test version\nIdle_pct: 100\n"
)

type haproxy struct {
	*httptest.Server
	response []byte
}

func newHaproxy(response []byte) *haproxy {
	h := &haproxy{response: response}
	h.Server = httptest.NewServer(handler(h))
	return h
}

func handler(h *haproxy) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Write(h.response)
	}
}

func handlerStale(exit chan bool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		<-exit
	}
}

func expectMetrics(t *testing.T, c prometheus.Collector, fixture string) {
	exp, err := os.Open(path.Join("test", fixture))
	if err != nil {
		t.Fatalf("Error opening fixture file %q: %v", fixture, err)
	}
	if err := testutil.CollectAndCompare(c, exp); err != nil {
		t.Fatal("Unexpected metrics returned:", err)
	}
}

func TestInvalidConfig(t *testing.T) {
	h := newHaproxy([]byte("not,enough,fields"))
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())

	expectMetrics(t, e, "invalid_config.metrics")
}

func TestServerWithoutChecks(t *testing.T) {
	h := newHaproxy([]byte("test,127.0.0.1:8080,0,0,0,0,0,0,0,0,,0,,0,0,0,0,no check,1,1,0,0,,,0,,1,1,1,,0,,2,0,,0,,,,0,0,0,0,0,0,0,,,,0,0,,,,,,,,,,,"))
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())

	expectMetrics(t, e, "server_without_checks.metrics")
}

// TestServerBrokenCSV ensures bugs in CSV format are handled gracefully. List of known bugs:
//
//   - http://permalink.gmane.org/gmane.comp.web.haproxy/26561
func TestServerBrokenCSV(t *testing.T) {
	const data = `foo,FRONTEND,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,0,,0,L4OK,,0,,,,,,,0,,,,0,0,,,,,,,,,,,
foo,bug-missing-comma,0,0,0,0,,0,0,0,,0,,0,0,0,0,DRAIN (agent)1,1,0,0,0,5007,0,,1,8,1,,0,,2,0,,0,L4OK,,0,,,,,,,0,,,,0,0,,,,,,,,,,,
foo,foo-instance-0,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,0,,0,L4OK,,0,,,,,,,0,,,,0,0,,,,,,,,,,,
foo,BACKEND,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,0,,0,L4OK,,0,,,,,,,0,,,,0,0,,,,,,,,,,,
`
	h := newHaproxy([]byte(data))
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())

	expectMetrics(t, e, "server_broken_csv.metrics")
}

func TestOlderHaproxyVersions(t *testing.T) {
	const data = `foo,FRONTEND,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,
foo,foo-instance-0,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,
foo,BACKEND,0,0,0,0,,0,0,0,,0,,0,0,0,0,UP,1,1,0,0,0,5007,0,,1,8,1,,0,,2,
`
	h := newHaproxy([]byte(data))
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())

	expectMetrics(t, e, "older_haproxy_versions.metrics")
}

func TestConfigChangeDetection(t *testing.T) {
	h := newHaproxy([]byte(""))
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())
	ch := make(chan prometheus.Metric)

	go func() {
		defer close(ch)
		e.Collect(ch)
	}()

	// TODO: Add a proper test here. Vet the possibilities of the new
	// client_golang to do this easily. If better test support is needed,
	// add it to client_golang first. (See also readCounter() above.)

	// Suck up the remaining metrics.
	for range ch {
	}
}

func TestDeadline(t *testing.T) {
	exit := make(chan bool)
	s := httptest.NewServer(handlerStale(exit))
	defer func() {
		// s.Close() will block until the handler
		// returns, so we need to make it exit.
		exit <- true
		s.Close()
	}()

	e, err := NewExporter(s.URL, true, false, serverMetrics, excludedServerStates, 1*time.Second, log.NewNopLogger())
	if err != nil {
		t.Fatal(err)
	}

	expectMetrics(t, e, "deadline.metrics")
}

func TestNotFound(t *testing.T) {
	s := httptest.NewServer(http.NotFoundHandler())
	defer s.Close()

	e, err := NewExporter(s.URL, true, false, serverMetrics, excludedServerStates, 1*time.Second, log.NewNopLogger())
	if err != nil {
		t.Fatal(err)
	}

	expectMetrics(t, e, "not_found.metrics")
}

func newHaproxyUnix(file, statsPayload string, infoPayload string) (io.Closer, error) {
	if err := os.Remove(file); err != nil && !os.IsNotExist(err) {
		return nil, err
	}
	l, err := net.Listen("unix", file)
	if err != nil {
		return nil, err
	}
	go func() {
		for {
			c, err := l.Accept()
			if err != nil {
				return
			}
			go func(c net.Conn) {
				defer c.Close()
				r := bufio.NewReader(c)
				for {
					l, err := r.ReadString('\n')
					if err != nil {
						return
					}
					switch l {
					case "show info\n":
						c.Write([]byte(infoPayload))
						return
					case "show stat\n":
						c.Write([]byte(statsPayload))
						return
					default:
						// invalid command
						return
					}
				}
			}(c)
		}
	}()
	return l, nil
}

func TestUnixDomain(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("not on windows")
		return
	}
	srv, err := newHaproxyUnix(testSocket, "test,127.0.0.1:8080,0,0,0,0,0,0,0,0,,0,,0,0,0,0,no check,1,1,0,0,,,0,,1,1,1,,0,,2,0,,0,,,,0,0,0,0,0,0,0,,,,0,0,,,,,,,,,,,\n", testInfo)
	if err != nil {
		t.Fatalf("can't start test server: %v", err)
	}
	defer srv.Close()

	e, err := NewExporter("unix:"+testSocket, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())
	if err != nil {
		t.Fatal(err)
	}

	expectMetrics(t, e, "unix_domain.metrics")
}

func TestUnixDomainNotFound(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("not on windows")
		return
	}

	if err := os.Remove(testSocket); err != nil && !os.IsNotExist(err) {
		t.Fatal(err)
	}
	e, _ := NewExporter("unix:"+testSocket, true, false, serverMetrics, excludedServerStates, 1*time.Second, log.NewNopLogger())
	expectMetrics(t, e, "unix_domain_not_found.metrics")
}

func TestUnixDomainDeadline(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("not on windows")
		return
	}

	exit := make(chan struct{})
	defer close(exit)

	if err := os.Remove(testSocket); err != nil && !os.IsNotExist(err) {
		t.Fatal(err)
	}
	l, err := net.Listen("unix", testSocket)
	if err != nil {
		t.Fatal(err)
	}
	defer l.Close()
	go func() {
		for {
			if _, err := l.Accept(); err != nil {
				return
			}
			go func() {
				// block
				<-exit
			}()
		}
	}()

	e, _ := NewExporter("unix:"+testSocket, true, false, serverMetrics, excludedServerStates, 1*time.Second, log.NewNopLogger())

	expectMetrics(t, e, "unix_domain_deadline.metrics")
}

func TestInvalidScheme(t *testing.T) {
	e, err := NewExporter("gopher://gopher.quux.org", true, false, serverMetrics, excludedServerStates, 1*time.Second, log.NewNopLogger())
	if expect, got := (*Exporter)(nil), e; expect != got {
		t.Errorf("expected %v, got %v", expect, got)
	}
	if err == nil {
		t.Fatalf("expected non-nil error")
	}
	if expect, got := err.Error(), `unsupported scheme: "gopher"`; expect != got {
		t.Errorf("expected %q, got %q", expect, got)
	}
}

func TestParseStatusField(t *testing.T) {
	tests := []struct {
		input string
		want  int64
	}{
		{"UP", 1},
		{"UP 1/3", 1},
		{"UP 2/3", 1},
		{"OPEN", 1},
		{"no check", 1},
		{"DOWN", 0},
		{"DOWN 1/2", 0},
		{"NOLB", 0},
		{"MAINT", 0}, // prometheus/haproxy_exporter#35
		{"unknown", 0},
	}

	for _, tt := range tests {
		if have := parseStatusField(tt.input); tt.want != have {
			t.Errorf("want status value %d for input %s, have %d",
				tt.want,
				tt.input,
				have,
			)
		}
	}
}

func TestFilterServerMetrics(t *testing.T) {
	tests := []struct {
		input string
		want  map[int]metricInfo
	}{
		{input: "", want: map[int]metricInfo{}},
		{input: "8", want: map[int]metricInfo{8: metricInfo{Desc: serverMetrics[8].Desc, Type: prometheus.CounterValue}}},
		{input: serverMetrics.String(), want: serverMetrics},
	}

	for _, tt := range tests {
		have, err := filterServerMetrics(tt.input)
		if err != nil {
			t.Errorf("unexpected error for input %s: %s", tt.input, err)
			continue
		}
		if !reflect.DeepEqual(tt.want, have) {
			t.Errorf("want filtered metrics %+v for input %q, have %+v",
				tt.want,
				tt.input,
				have,
			)
		}
	}
}

func BenchmarkExtract(b *testing.B) {
	config, err := os.ReadFile("test/haproxy.csv")
	if err != nil {
		b.Fatalf("could not read config file: %v", err.Error())
	}

	h := newHaproxy(config)
	defer h.Close()

	e, _ := NewExporter(h.URL, true, false, serverMetrics, excludedServerStates, 5*time.Second, log.NewNopLogger())

	var before, after runtime.MemStats
	runtime.GC()
	runtime.ReadMemStats(&before)
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		ch := make(chan prometheus.Metric)
		go func(ch chan prometheus.Metric) {
			for range ch {
			}
		}(ch)

		e.Collect(ch)
		close(ch)
	}

	runtime.GC()
	runtime.ReadMemStats(&after)

	b.Logf("%d bytes used after %d runs", after.Alloc-before.Alloc, b.N)
}


proxy.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/log"
)

type ValSetter interface {
	Set(float64)
}

type MetricVecProxy struct {
	mv interface{}
}

func NewMetricVecProxy(family MetricFamily, opts prometheus.Opts, labelNames []string) *MetricVecProxy {
	switch family {
	case Counter:
		return &MetricVecProxy{
			mv: prometheus.NewCounterVec(
				prometheus.CounterOpts{
					Namespace:   opts.Namespace,
					Name:        opts.Name,
					Help:        opts.Help,
					ConstLabels: opts.ConstLabels,
				},
				labelNames,
			),
		}
	case Gauge:
		return &MetricVecProxy{
			mv: prometheus.NewGaugeVec(
				prometheus.GaugeOpts{
					Namespace:   opts.Namespace,
					Name:        opts.Name,
					Help:        opts.Help,
					ConstLabels: opts.ConstLabels,
				},
				labelNames,
			),
		}
	default:
		log.Fatal("Unsupported metric family: %T", family)
	}

	return nil
}

func (p *MetricVecProxy) Collect(ch chan<- prometheus.Metric) {
	switch v := p.mv.(type) {
	case *prometheus.CounterVec:
		v.Collect(ch)
	case *prometheus.GaugeVec:
		v.Collect(ch)
	default:
		log.Fatal("Collect: Unsupported metric vector type: %T", v)
	}
}

func (p *MetricVecProxy) Describe(ch chan<- *prometheus.Desc) {
	switch v := p.mv.(type) {
	case *prometheus.CounterVec:
		v.Describe(ch)
	case *prometheus.GaugeVec:
		v.Describe(ch)
	default:
		log.Fatal("Describe: unsupported metric vector type: %T", v)
	}
}

func (p *MetricVecProxy) Reset() {
	switch v := p.mv.(type) {
	case *prometheus.CounterVec:
		v.Reset()
	case *prometheus.GaugeVec:
		v.Reset()
	default:
		log.Fatal("Reset: unsupported metric vector type: %T", v)
	}
}

func (p *MetricVecProxy) WithLabelValues(lvs ...string) ValSetter {
	switch v := p.mv.(type) {
	case *prometheus.CounterVec:
		return v.WithLabelValues(lvs...)
	case *prometheus.GaugeVec:
		return v.WithLabelValues(lvs...)
	default:
		log.Fatal("WithLabelValues: unsupported metric type: %T", v)
	}

	return nil
}


metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
)

type MetricFamily int

const (
	Counter MetricFamily = iota
	Gauge
)

// HAProxy 1.4
// pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,
// HAProxy 1.5
// pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,comp_in,comp_out,comp_byp,comp_rsp,lastsess,
// HAProxy 1.6
// pxname,svname,qcur,qmax,scur,smax,slim,stot,bin,bout,dreq,dresp,ereq,econ,eresp,wretr,wredis,status,weight,act,bck,chkfail,chkdown,lastchg,downtime,qlimit,pid,iid,sid,throttle,lbtot,tracked,type,rate,rate_lim,rate_max,check_status,check_code,check_duration,hrsp_1xx,hrsp_2xx,hrsp_3xx,hrsp_4xx,hrsp_5xx,hrsp_other,hanafail,req_rate,req_rate_max,req_tot,cli_abrt,srv_abrt,comp_in,comp_out,comp_byp,comp_rsp,lastsess,last_chk,last_agt,qtime,ctime,rtime,ttime,

const (
	namespace             = "haproxy" // For Prometheus metrics.
	expectedCsvFieldCount = 52
	statusField           = 17
)

type metricSpec struct {
	metricName   string
	docString    string
	metricFamily MetricFamily
	metricLabels prometheus.Labels
}

var (
	metricSpecs = []metricSpec{
		// 0
		{},
		{},
		{
			// qcur
			"current_queue",
			"Current number of queued requests assigned to this server",
			Gauge,
			nil,
		},
		{
			// qmax
			"max_queue",
			"Maximum observed number of queued requests assigned to this server.",
			Gauge,
			nil,
		},
		{
			// scur
			"current_sessions",
			"Current number of active sessions.",
			Gauge,
			nil,
		},

		// 5
		{
			// smax
			"max_sessions",
			"Maximum observed number of active sessions.",
			Gauge,
			nil,
		},
		{
			// slim
			"limit_sessions",
			"Configured session limit.",
			Gauge,
			nil,
		},
		{
			// stot
			"connections_total",
			"Total number of connections.",
			Counter,
			nil,
		},
		{
			// bin
			"bytes_in_total",
			"Current total of incoming bytes.",
			Counter,
			nil,
		},
		{
			// bout
			"bytes_out_total",
			"Current total of outgoing bytes.",
			Counter,
			nil,
		},

		// 10
		{
			// dreq
			"requests_denied_total",
			"Total of requests denied for security.",
			Counter,
			nil,
		},
		{
		// dresp
		},
		{
			// ereq
			"request_errors_total",
			"Total of request errors.",
			Counter,
			nil,
		},
		{
			// econ
			"connection_errors_total",
			"Total of connection errors.",
			Counter,
			nil,
		},
		{
			// eresp
			"response_errors_total",
			"Total of response errors.",
			Counter,
			nil,
		},

		// 15
		{
			// wretr
			"retry_warnings_total",
			"Total of retry warnings.",
			Counter,
			nil,
		},
		{
			// wredis
			"redispatch_warnings_total",
			"Total of redispatch warnings.",
			Counter,
			nil,
		},
		{
			// status
			"up",
			"Current health status of the server (1 = UP, 0 = DOWN }.",
			Gauge,
			nil,
		},
		{
			// weight
			"weight",
			"Current weight of the server.",
			Gauge,
			nil,
		},
		{
		// act
		},

		// 20
		{
		// bck
		},
		{
			// chkfail
			"check_failures_total",
			"Total number of failed health checks.",
			Counter,
			nil,
		},
		{
		// chkdown
		},
		{
		// lastchg
		},
		{
			// downtime
			"downtime_seconds_total",
			"Total downtime in seconds.",
			Counter,
			nil,
		},

		// 25
		{
		// qlimit
		},
		{
		// pid
		},
		{
		// iid
		},
		{
		// sid
		},
		{
		// throttle
		},

		// 30
		{
		// lbtot
		},
		{
		// tracked
		},
		{
		// type
		},
		{
			// rate
			"current_session_rate",
			"Current number of sessions per second over last elapsed second.",
			Gauge,
			nil,
		},
		{
			// rate_lim
			"limit_session_rate",
			"Configured limit on new sessions per second.",
			Gauge,
			nil,
		},

		// 35
		{
			// rate_max
			"max_session_rate",
			"Maximum observed number of sessions per second.",
			Gauge,
			nil,
		},
		{
		// check_status
		},
		{
		// check_code
		},
		{
			// check_duration
			"check_duration_milliseconds",
			"Previously run health check duration, in milliseconds",
			Gauge,
			nil,
		},
		{
			// hrsp_1xx
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "1xx"},
		},

		// 40
		{
			// hrsp_2xx
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "2xx"},
		},
		{
			// hrsp_3xx
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "3xx"},
		},
		{
			// hrsp_4xx
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "4xx"},
		},
		{
			// hrsp_5xx
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "5xx"},
		},
		{
			// hrsp_other
			"http_responses_total",
			"Total of HTTP responses.",
			Counter,
			prometheus.Labels{"code": "other"},
		},

		// 45
		{
		// hanafail
		},
		{
		// req_rate
		},
		{
		// req_rate_max
		},
		{
			// req_tot
			"http_requests_total",
			"Total HTTP requests.",
			Counter,
			nil,
		},
		{
		// cli_abrt
		},

		// 50
		{
		// srv_abrt
		},
		{
		// comp_in
		},
		{
		// comp_out
		},
		{
		// comp_byp
		},
		{
		// comp_rsp
		},

		// 55
		{
		// lastsess
		},
		{
		// last_chk
		},
		{
		// last_agt
		},
		{
		// qtime
		},
		{
		// ctime
		},

		// 60
		{
		// rtime
		},
		{
		// ttime
		},
	}
)
